# How-to-perfom-a-MWAS-at-Ayixon-Lab
This repository intens to serve as a blueprint to perform a microbiome-wide association study on the archezoa server.
The main software useded is Pyseer (Lees et al., 2017), how ever others are also used, if not mentioned all are is installed on $PATH.
For further Pyseer information refer to : https://pyseer.readthedocs.io/en/master/tutorial.html 

## Prepare the input files

We'll assume that all assemblies are on a folder, hereinafter referred to as "working_folder".
* Pro Tip: When working with several files and it they have a binary clasificaition e.g. Positive/Negative
  I recommed to rename according each sample according to its classification, that will help to keep a track 
First, a "phenotype" file is requiere. This is a tab separeted file with two columns, header requiere.

```
```
## Variant (unitigs) Calling
```
```
## Desining an Kinship Matrix 
```
```
## Performing and linear mixed model on Pysser.
```
```
## Variant annotation 
# References
Lees, John A., Galardini, M., et al. pyseer: a comprehensive tool for microbial pangenome-wide association studies. 
Bioinformatics 34:4310â€“4312 (2018). doi:10.1093/bioinformatics/bty539. 

